### PR TITLE
Feature/consensus integration to master

### PR DESCRIPTION
**Pull request by n-hutton**
Latest consensus changes plus a few bugfixes. If you would like to know how consensus works in detail I am happy to discuss. In short, the blockchain will contain a structure, the 'block entropy' which is a continuous stream of group signatures by various cabinets (a collection of nodes). That is, each block contains a signature of the previous block entropy signature.

Since the signatures are random/white, this can be used on a block by block basis to determine the priority of 'who should produce a block' - which is a deterministic shuffle of the valid miners at that point in time.

The major output from this is that the genesis file must be set up correctly for the nodes in the network, as there must always be stakers/miners in the system.

### CODE REVIEW

File: libs/core/include/core/reactor.hpp [hunk 1]
```diff
@@ -40,7 +40,8 @@ class Reactor
   Reactor(Reactor &&)      = delete;
   ~Reactor()               = default;
 
-  bool Attach(WeakRunnable runnable);
+  bool Attach(WeakRunnable runnable);   // NOLINT
```
Comment (n-hutton) [inline libs/core/include/core/reactor.hpp:43]: Note here addition of attaching a vector of runnables

File: libs/ledger/include/ledger/chain/block.hpp [hunk 2]
```diff
@@ -42,36 +43,40 @@ namespace ledger {
...
+    Digest       hash;               ///< The hash of the block
+    Digest       previous_hash;      ///< The hash of the previous block
+    Digest       merkle_hash;        ///< The merkle state hash across all shards
+    uint64_t     block_number{0};    ///< The height of the block from genesis
+    Address      miner;              ///< The identity of the generated miner
+    uint32_t     log2_num_lanes{0};  ///< The log2(number of lanes)
+    Slices       slices;             ///< The slice lists
+    DAGEpoch     dag_epoch;          ///< DAG epoch containing information on new dag_nodes
+    uint64_t     timestamp{0u};      ///< The number of seconds elapsed since the Unix epoch
+    BlockEntropy block_entropy;      ///< Entropy that determines miner priority for the next block
```
Comment (n-hutton) [inline libs/ledger/include/ledger/chain/block.hpp:69]: chain now contains block entropy

File: libs/ledger/include/ledger/protocols/main_chain_rpc_protocol.hpp [hunk 3]
```diff
@@ -55,6 +55,7 @@ class MainChainProtocol : public service::Protocol
   {
     MainChain::Blocks blocks;
 
+    // TODO(issue 1725): this can cause issue if it doesn't exist (?)
```
Comment (n-hutton) [inline libs/ledger/include/ledger/protocols/main_chain_rpc_protocol.hpp:58]: unrelated issue

File: libs/muddle/internal/muddle_fake.hpp [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+
+/**
+ * Construct a global in-memory network which can be used by fake muddles.
+ *
+ * To do this, create a map of address to a collection of packets. Peers can then 'send'
+ * packets by writing into this collection. If the receiving muddle is listening, it has
+ * a responsibility to collect the packets.
+ *
+ */
+class FakeNetwork
```
Comment (n-hutton) [inline libs/muddle/internal/muddle_fake.hpp:36]: For testing, a 'fake muddle' + network is created. Will allow network-esque tests to be run in parallel without port limitations.

File: libs/muddle/src/router.cpp [hunk 5]
```diff
@@ -831,6 +831,12 @@ void Router::RoutePacket(PacketPtr const &packet, bool external)
       return;
     }
 
+    if (kademlia_routing_)
```
Comment (n-hutton) [inline libs/muddle/src/router.cpp:834]: Ed's kademlia routing changes - ask him

File: apps/beacon/main.cpp [hunk 6]
```diff
@@ -180,19 +181,20 @@ int main()
       {
         member->beacon_service.StartNewCabinet(cabinet, static_cast<uint32_t>(cabinet.size() / 2),
                                                block_number, block_number + aeon_length,
-                                               static_cast<uint64_t>(std::time(nullptr)));
+                                               static_cast<uint64_t>(std::time(nullptr)),
```
Comment (ejfitzgerald) [inline apps/beacon/main.cpp:184]: Not in the scope of this PR but we should add a wrapper around this I think and update all the parts of the code base (this part being the UTC clock getter)

File: libs/beacon/benchmark/entropy_gen_bench.cpp [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+{
+public:
+  BeaconServiceInsertable(MuddleInterface &muddle, ledger::ManifestCacheInterface &manifest_cache,
+                          CertificatePtr certificate, SharedEventManager event_manager)  // NOLINT
+    : BeaconService(muddle, manifest_cache, certificate, std::move(event_manager))
+  {}
+
+  void PushNewExecUnit(const SharedAeonExecutionUnit &beacon)
+  {
+    // TODO(HUT): consider whether to lock here for the test.
```
Comment (ejfitzgerald) [inline libs/beacon/benchmark/entropy_gen_bench.cpp:67]: Should we resolve this now?

File: libs/beacon/examples/beacon_manager/main.cpp [hunk 8]
```diff
@@ -61,7 +61,7 @@ int main(int argc, char **argv)
   for (auto &m : nodes)
   {
     auto &n = *m.second;
-    participants.emplace_back(n.identity(), n.id());
+    participants.push_back({n.identity(), n.id()});
```
Comment (ejfitzgerald) [inline libs/beacon/examples/beacon_manager/main.cpp:64]: What prompted this change?

File: libs/beacon/include/beacon/aeon.hpp [hunk 9]
```diff
@@ -32,29 +34,33 @@ namespace beacon {
...
+  std::set<MuddleAddress> members{};
+  uint64_t                round_start{0};
+  uint64_t                round_end{0};
+  BlockEntropy            block_entropy_previous;
 
   // Timeouts for waiting for other members
   uint64_t start_reference_timepoint{uint64_t(-1)};
 };
 
+// TODO(HUT): merge these into just Aeon
```
Comment (ejfitzgerald) [inline libs/beacon/include/beacon/aeon.hpp:51]: Issue?

File: libs/beacon/src/beacon_service.cpp [hunk 10]
```diff
@@ -24,113 +24,56 @@
...
   , endpoint_{muddle.GetEndpoint()}
   , state_machine_{std::make_shared<StateMachine>("BeaconService", State::WAIT_FOR_SETUP_COMPLETION,
                                                   ToString)}
-  , blocks_per_round_{blocks_per_round}
-  , group_public_key_subscription_{endpoint_.Subscribe(SERVICE_DKG, CHANNEL_PUBLIC_KEY)}
-  , entropy_subscription_{endpoint_.Subscribe(SERVICE_DKG, CHANNEL_ENTROPY_DISTRIBUTION)}
   , rpc_client_{"BeaconService", endpoint_, SERVICE_DKG, CHANNEL_RPC}
   , event_manager_{std::move(event_manager)}
-  , cabinet_creator_{muddle, identity_, manifest_cache}  // TODO(tfr): Make shared
+  , cabinet_creator_{muddle, identity_, manifest_cache, certificate_}  // TODO(tfr): Make shared
```
Comment (ejfitzgerald) [inline libs/beacon/src/beacon_service.cpp:49]: Remove todo?

File: libs/beacon/src/beacon_service.cpp [hunk 11]
```diff
@@ -484,99 +270,93 @@ BeaconService::State BeaconService::OnVerifySignaturesState()
...
     }
 
-    FETCH_LOG_INFO(LOGGING_NAME, "Generated new entropy value");
+    FETCH_LOG_INFO(LOGGING_NAME, "After adding, we have ", all_sigs_map.size(),
+                   " signatures. Round: ", index);
+  }  // Mutex unlocks here since verification can take some time
+
+  MilliTimer const timer{"Verify threshold signature", 100};
 
+  // TODO(HUT): possibility for infinite loop here I suspect.
```
Comment (ejfitzgerald) [inline libs/beacon/src/beacon_service.cpp:321]: Issue?

File: libs/beacon/src/beacon_setup_service.cpp [hunk 12]
```diff
@@ -86,48 +91,47 @@ uint64_t GetTime()
   return static_cast<uint64_t>(std::time(nullptr));
 }
 
+//#define CHANNEL_TYPE RBC
+#define CHANNEL_TYPE muddle::RBC
```
Comment (ejfitzgerald) [inline libs/beacon/src/beacon_setup_service.cpp:95]: This should be refactoring into a factory

File: libs/beacon/src/beacon_setup_service.cpp [hunk 13]
```diff
@@ -669,113 +667,101 @@ BeaconSetupService::State BeaconSetupService::OnWaitForReconstructionShares()
...
- * group public signature
+ * group public signature. If successful, this should generate the first block entropy
+ * structure.
+ *
+ * To do this, a block entropy struct specifying the qualified members, group signature etc. is
+ * created, and nodes try and collect threshold signatures (personal) of that hash from members of
+ * the qualified set.
+ *
  */
+// TODO(HUT): rename dry run to create first signature.
```
Comment (ejfitzgerald) [inline libs/beacon/src/beacon_setup_service.cpp:703]: Issue?

File: libs/beacon/src/beacon_setup_service.cpp [hunk 14]
```diff
@@ -669,113 +667,101 @@ BeaconSetupService::State BeaconSetupService::OnWaitForReconstructionShares()
...
 BeaconSetupService::State BeaconSetupService::OnDryRun()
 {
-  std::lock_guard<std::mutex> lock(mutex_);
+  FETCH_LOCK(mutex_);
+
+  // TODO(HUT): reset to qual here for the networking (?)
```
Comment (ejfitzgerald) [inline libs/beacon/src/beacon_setup_service.cpp:708]: Issue?

File: libs/network/src/tcp/tcp_server.cpp [hunk 15]
```diff
@@ -133,7 +133,7 @@ void TCPServer::Start()
 
 void TCPServer::Stop()
 {
-  // TODO(EJF): Hmmmmmm...?
+  // TODO(EJF): Hmmmmmm...???
```
Comment (ejfitzgerald) [inline libs/network/src/tcp/tcp_server.cpp:136]: I guess revert

File: libs/muddle/src/router.cpp (hunk 5 above)
Comment (ejfitzgerald) [reply libs/muddle/src/router.cpp:834]: This is a duplicate of the line above and they are both wrong. I will fix this in a subsequent PR (along with additional updates)

File: libs/muddle/include/muddle/rbc_messages.hpp [hunk 16]
```diff
@@ -234,8 +234,8 @@ struct MapSerializer<muddle::RBCMessage, D>
     uint8_t type;
     map.ExpectKeyGetValue(TYPE, type);
     map.ExpectKeyGetValue(CHANNEL, msg.channel_);
-    // TODO(?): Remove and deduce from network connection
-    map.ExpectKeyGetValue(ADDRESS, msg.id_);
+    map.ExpectKeyGetValue(ADDRESS,
+                          msg.id_);  // TODO (ed): Remove and deduce from network connection
```
Comment (ejfitzgerald) [inline libs/muddle/include/muddle/rbc_messages.hpp:238]: I can add to my TODO list

Comment (ejfitzgerald) [reply libs/muddle/include/muddle/rbc_messages.hpp:238]: Lets raise an issue to track

File: libs/ledger/tests/chain/block_coordinator_tests.cpp [hunk 17]
```diff
@@ -990,6 +990,18 @@ class NiceMockBlockCoordinatorTests : public BlockCoordinatorTests
     execution_manager_ = std::make_unique<NiceMock<MockExecutionManager>>(storage_unit_->fake);
     packer_            = std::make_unique<NiceMock<MockBlockPacker>>();
     block_sink_        = std::make_unique<FakeBlockSink>();
+
+    // BlockCoordinator(MainChain &chain, DAGPtr dag, ExecutionManagerInterface &execution_manager,
```
Comment (ejfitzgerald) [inline libs/ledger/tests/chain/block_coordinator_tests.cpp:994]: Remove?

File: libs/ledger/include/ledger/protocols/main_chain_rpc_protocol.hpp (hunk 3 above)
Comment (ejfitzgerald) [reply libs/ledger/include/ledger/protocols/main_chain_rpc_protocol.hpp:58]: Remove?

**Review by ejfitzgerald — APPROVED** [review]
I am approving since I think this is good enough to merge in its current state. There are however a number of touch up items that I have pointed out. Also for the sake of the build time we need to move the definitions of most of the methods to corresponding CPP files

File: libs/ledger/include/ledger/protocols/main_chain_rpc_protocol.hpp (hunk 3 above)
Comment (n-hutton) [reply libs/ledger/include/ledger/protocols/main_chain_rpc_protocol.hpp:58]: Not removing as per discussion

File: apps/beacon/main.cpp (hunk 6 above)
Comment (n-hutton) [reply apps/beacon/main.cpp:184]: Todo noted

File: libs/beacon/benchmark/entropy_gen_bench.cpp (hunk 7 above)
Comment (n-hutton) [reply libs/beacon/benchmark/entropy_gen_bench.cpp:67]: Jenny is/has resolved it

File: libs/beacon/examples/beacon_manager/main.cpp (hunk 8 above)
Comment (n-hutton) [reply libs/beacon/examples/beacon_manager/main.cpp:64]: a bad merge - reverting now

File: libs/beacon/include/beacon/aeon.hpp (hunk 9 above)
Comment (n-hutton) [reply libs/beacon/include/beacon/aeon.hpp:51]: TODO number 2

File: libs/beacon/src/beacon_service.cpp (hunk 10 above)
Comment (n-hutton) [reply libs/beacon/src/beacon_service.cpp:49]: OK

File: libs/beacon/src/beacon_service.cpp (hunk 11 above)
Comment (n-hutton) [reply libs/beacon/src/beacon_service.cpp:321]: TODO 3 - partially jenny is workin on this

File: libs/beacon/src/beacon_setup_service.cpp (hunk 12 above)
Comment (n-hutton) [reply libs/beacon/src/beacon_setup_service.cpp:95]: OK

File: libs/beacon/src/beacon_setup_service.cpp (hunk 13 above)
Comment (n-hutton) [reply libs/beacon/src/beacon_setup_service.cpp:703]: todo 4

File: libs/beacon/src/beacon_setup_service.cpp (hunk 14 above)
Comment (n-hutton) [reply libs/beacon/src/beacon_setup_service.cpp:708]: Need to ask Jenny about this

**Pull request merged**